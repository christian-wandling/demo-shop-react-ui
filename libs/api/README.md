# api

A comprehensive library for consuming the demo-shop api. Supports accessing a product catalog, user management,
shopping cart operations, and order processing.
