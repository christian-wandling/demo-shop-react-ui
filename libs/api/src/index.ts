export * from './lib/openapi';
export * from './lib/ApiContext';
export * from './lib/services/apiService';
export * from './lib/ApiProvider';
export * from './lib/services/apiService';
export * from './lib/middleware/authMiddleware';
