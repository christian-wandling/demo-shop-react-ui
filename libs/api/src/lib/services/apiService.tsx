import { ApiContextType } from '../models/apiContextType';
import { Configuration, ConfigurationParameters, OrderApi, ProductApi, ShoppingSessionApi, UserApi } from '../openapi';

let apiInstances: ApiContextType | null = null;

export const initApi = (apiConfig: ConfigurationParameters): ApiContextType => {
  if (apiInstances) return apiInstances;

  const config = new Configuration(apiConfig);

  apiInstances = {
    orderApi: new OrderApi(config),
    productApi: new ProductApi(config),
    shoppingSessionApi: new ShoppingSessionApi(config),
    userApi: new UserApi(config),
  };

  return apiInstances;
};

export const getApi = (): ApiContextType => {
  if (!apiInstances) {
    throw new Error('API service is not initialized. Call `createApis()` first.');
  }
  return apiInstances;
};
