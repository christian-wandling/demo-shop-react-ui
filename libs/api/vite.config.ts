import { defineConfig } from 'vite';
import react from '@vitejs/plugin-react';

export default defineConfig(() => ({
  root: __dirname,
  cacheDir: '../../node_modules/.vite/libs/api',
  plugins: [react()],
  // Uncomment this if you are using workers.
  // worker: {
  //  plugins: [ nxViteTsPaths() ],
  // },
  test: {
    watch: false,
    globals: true,
    environment: 'jsdom',
    reporters: ['default'],
    passWithNoTests: true,
    coverage: {
      reportsDirectory: './test-output/vitest/coverage',
      enabled: false,
    },
  },
}));
