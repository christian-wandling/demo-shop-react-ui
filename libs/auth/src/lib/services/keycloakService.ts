import Keycloak from 'keycloak-js';
import { AuthConfig } from '../models/authConfig';
import { PermissionStrategy } from '../enums/permissionStrategy';
import { PERMISSION_STRATEGIES } from '../constants/permissionStrategies';

let keycloak: Keycloak | null = null;

/**
 * Initializes Keycloak instance (singleton)
 */
export function initKeycloak(authConfig: AuthConfig): Promise<boolean> {
  if (keycloak) {
    return Promise.resolve(true);
  }

  keycloak = new Keycloak(authConfig.keycloak);

  return keycloak.init({
    onLoad: 'check-sso',
    silentCheckSsoRedirectUri: window.location.origin + '/silent-check-sso.html',
  });
}

/**
 * Returns the Keycloak token
 */
export function getToken(): string | undefined {
  return keycloak?.token;
}

/**
 * Checks if user is authenticated
 */
export function isAuthenticated(): boolean {
  return keycloak?.authenticated === true;
}

/**
 * Logs in the user
 */
export function login(): Promise<void> {
  return keycloak?.login() ?? Promise.resolve();
}

/**
 * Logs out the user
 */
export function logout(): Promise<void> {
  return keycloak?.logout({ redirectUri: window.location.origin }) ?? Promise.resolve();
}

/**
 * Registers a new user
 */
export function register(): Promise<void> {
  return keycloak?.register() ?? Promise.resolve();
}

/**
 * Utility function to check for permissions
 */
export function hasPermission(permissionStrategy: PermissionStrategy, ...args: unknown[]): boolean {
  const permissionStrategyFn = PERMISSION_STRATEGIES[permissionStrategy];

  if (!permissionStrategyFn) {
    return false;
  }

  return permissionStrategyFn();
}
