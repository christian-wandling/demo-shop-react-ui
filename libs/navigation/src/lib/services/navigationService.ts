// Assuming this exists in React form
import { NavigationItem } from '../models/navigationItem';
import { hasPermission, PermissionStrategy } from '@demo-shop-react-ui/auth';
import { NavigationItemType } from '../enums/navigationItemType';

/**
 * Returns navigation items filtered by type and user permissions
 *
 * This method filters the navigation items based on:
 * 1. The specified navigation type
 * 2. The user's permissions (if a permission strategy is defined)
 * The resulting items are sorted by their order property in ascending order.
 */
export function getFilteredNavigationItems(
  items: NavigationItem[],
  type: NavigationItemType = NavigationItemType.ROUTE
): NavigationItem[] {
  const isType = (itemType: NavigationItemType) => itemType === type;
  const allowDisplay = (strategy?: PermissionStrategy) => !strategy || hasPermission(strategy);

  return items
    .filter(item => isType(item.type) && allowDisplay(item.options?.['permissionStrategy']))
    .sort((a, b) => a.order - b.order);
}
