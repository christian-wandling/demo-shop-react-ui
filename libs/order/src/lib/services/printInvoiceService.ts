import { jsPDF } from 'jspdf';
import { OrderResponse, UserResponse } from '@demo-shop-react-ui/api';
import { format } from 'date-fns/format';
import { CompanyData } from '../models/company-data';
import { PaymentTerms } from '../models/payment-terms';
import { DrawOptions } from '../models/draw-options';
import { PdfTableColumns } from '../models/pdf-table.columns';
import { columns, companyData, createDrawOptions, paymentTerms } from '../constants/print-invoice.config';

/**
 * Generates a complete PDF invoice document for a given order
 *
 * Creates a PDF document with order details including company information, logo,
 * customer details, invoice number, date, and itemized order contents.
 *
 * @param order - Order information containing items and pricing details
 * @param user - User information for the customer who placed the order
 */
export function generatePdf(order: OrderResponse, user: UserResponse): void {
  if (!order) {
    throw new Error('Error printing invoice, order is missing');
  }

  if (!user) {
    throw new Error('Error printing invoice, user is missing');
  }

  const doc = new jsPDF('p', 'mm', 'A4', true);
  const drawOptions = createDrawOptions(doc.internal.pageSize.getWidth(), doc.internal.pageSize.getHeight());

  addFooter(doc, drawOptions, paymentTerms);
  addLogo(doc, drawOptions);
  addCompanyInformation(doc, drawOptions, companyData);
  addInvoiceDetails(doc, drawOptions, order);
  addCustomerInformation(doc, drawOptions, user);
  addTableHeader(doc, drawOptions, columns);
  addTableItems(doc, drawOptions, columns, order, paymentTerms);
  addTableFooter(doc, drawOptions, columns, order);

  doc.save(`invoice-${order.id}.pdf`);
}

/**
 * Adds the company logo to the PDF document
 *
 * Positions and inserts the logo image at the top-left corner of the document.
 *
 * @param doc - The jsPDF document instance
 * @param options - Drawing options with positioning information
 */
function addLogo(doc: jsPDF, options: DrawOptions): void {
  options.pointer.x = options.border.left;
  options.pointer.y = options.border.top;

  const width = 25;
  const height = 25;

  doc.addImage('icons/demo-shop.png', 'PNG', options.pointer.x, options.pointer.y, width, height);
}

/**
 * Adds company information to the PDF document
 *
 * Displays company details including name, address, contact information
 * at the top-right corner of the document.
 *
 * @param doc - The jsPDF document instance
 * @param options - Drawing options with positioning information
 * @param companyData - Company information to display
 */
function addCompanyInformation(doc: jsPDF, options: DrawOptions, companyData: CompanyData): void {
  options.pointer.x = options.border.right;
  options.pointer.y += 6;
  doc.setFont(options.text.font.default, 'bold');
  doc.setFontSize(options.text.size.small);
  doc.setTextColor(options.text.color.light);
  doc.text(companyData.name, options.pointer.x, options.pointer.y, { align: 'right' });

  options.pointer.y += 4;
  doc.setFont(options.text.font.default);
  doc.setFontSize(options.text.size.smaller);
  doc.text(companyData.address, options.pointer.x, options.pointer.y, { align: 'right' });

  options.pointer.y += 4;
  doc.text(companyData.city + ', ' + companyData.zipCode, options.pointer.x, options.pointer.y, { align: 'right' });

  options.pointer.y += 4;
  doc.text(companyData.phone, options.pointer.x, options.pointer.y, { align: 'right' });

  options.pointer.y += 4;
  doc.text(companyData.email, options.pointer.x, options.pointer.y, { align: 'right' });
}

/**
 * Adds invoice details to the PDF document
 *
 * Displays the invoice number and creation date in the document.
 *
 * @param doc - The jsPDF document instance
 * @param options - Drawing options with positioning information
 * @param order - Order information containing the invoice ID and creation date
 */
function addInvoiceDetails(doc: jsPDF, options: DrawOptions, order: OrderResponse): void {
  options.pointer.y += 22;
  doc.setFontSize(options.text.size.smaller);
  doc.setFont(options.text.font.default, 'bold');
  doc.text(`INVOICE #${order.id}`, options.pointer.x, options.pointer.y, { align: 'right' });

  options.pointer.y += 4;
  const date = format(order.created, 'yyyy-MM-dd');
  doc.setFont(options.text.font.default);
  doc.text(date, options.pointer.x, options.pointer.y, { align: 'right' });
}

/**
 * Adds customer information to the PDF document
 *
 * Displays customer details including name, address, and contact information.
 *
 * @param doc - The jsPDF document instance
 * @param options - Drawing options with positioning information
 * @param user - User information for the customer
 */
function addCustomerInformation(doc: jsPDF, options: DrawOptions, user: UserResponse): void {
  options.pointer.x = options.border.left;
  options.pointer.y += 15;
  doc.setFontSize(options.text.size.smaller);
  doc.setFont('Helvetica', 'bold');
  doc.text('YOUR DATA:', options.pointer.x, options.pointer.y);
  doc.setFont('Helvetica');

  options.pointer.y += 6;
  doc.text(`${user.firstname} ${user.lastname}`, options.pointer.x, options.pointer.y);

  options.pointer.y += 4;
  doc.text(`${user.address?.street} ${user.address?.apartment}`, options.pointer.x, options.pointer.y);

  options.pointer.y += 4;
  doc.text(`${user.address?.city}, ${user.address?.zip}`, options.pointer.x, options.pointer.y);

  options.pointer.y += 4;
  doc.text(user.address?.country ?? '', options.pointer.x, options.pointer.y);

  options.pointer.y += 4;
  doc.text(user.email, options.pointer.x, options.pointer.y);
}

/**
 * Adds the table header for order items to the PDF document
 *
 * Creates column headers for article, quantity, price, and amount.
 *
 * @param doc - The jsPDF document instance
 * @param options - Drawing options with positioning information
 * @param columns - Column configuration for the table
 */
function addTableHeader(doc: jsPDF, options: DrawOptions, columns: PdfTableColumns): void {
  options.pointer.y += 20;
  addLine(doc, options);

  options.pointer.y += 4;
  doc.setFontSize(options.text.size.smaller);
  doc.setTextColor(options.text.color.darker);
  doc.setFont('Helvetica', 'bold');

  doc.text(columns.article.label, columns.article.x, options.pointer.y, { baseline: 'middle' });
  doc.text(columns.qty.label, columns.qty.x, options.pointer.y, { baseline: 'middle', align: 'center' });
  doc.text(columns.price.label, columns.price.x, options.pointer.y, { baseline: 'middle', align: 'right' });
  doc.text(columns.amount.label, columns.amount.x, options.pointer.y, { baseline: 'middle', align: 'right' });

  options.pointer.y += 4;
  addLine(doc, options);
}

/**
 * Adds order item details to the PDF document
 *
 * Creates rows for each item in the order with product name, quantity,
 * unit price, and total price. Handles automatic page breaks if needed.
 *
 * @param doc - The jsPDF document instance
 * @param options - Drawing options with positioning information
 * @param columns - Column configuration for the table
 * @param order - Order information containing the items
 * @param paymentTerms - Payment terms to display if a new page is created
 */
function addTableItems(
  doc: jsPDF,
  options: DrawOptions,
  columns: PdfTableColumns,
  order: OrderResponse,
  paymentTerms: PaymentTerms
): void {
  const limitForNewContent = options.border.bottom - 40;

  options.pointer.y += 4;
  doc.setFont('Helvetica', 'normal');

  order.items.forEach(item => {
    options.pointer.y += 6;

    if (options.pointer.y > limitForNewContent) {
      addPage(doc, options, paymentTerms);
    }

    doc.setTextColor(options.text.color.dark);
    doc.setFontSize(options.text.size.smaller);
    doc.text(item.productName, columns.article.x, options.pointer.y);
    doc.text(item.quantity.toString(), columns.qty.x, options.pointer.y, { align: 'center' });
    doc.text(`$${item.unitPrice.toFixed(2)}`, columns.price.x, options.pointer.y, { align: 'right' });
    doc.text(`$${item.totalPrice.toFixed(2)}`, columns.amount.x, options.pointer.y, { align: 'right' });
  });
}

/**
 * Adds table footer with order totals to the PDF document
 *
 * Displays the total amount for the entire order.
 *
 * @param doc - The jsPDF document instance
 * @param options - Drawing options with positioning information
 * @param columns - Column configuration for the table
 * @param order - Order information containing the total amount
 */
function addTableFooter(doc: jsPDF, options: DrawOptions, columns: PdfTableColumns, order: OrderResponse): void {
  options.pointer.y += 6;
  addLine(doc, options);

  options.pointer.y += 6;
  doc.setFont('Helvetica', 'bold');
  doc.setFontSize(options.text.size.smaller);
  doc.text('Total', columns.price.x, options.pointer.y, { align: 'right' });
  doc.text(`$${order.amount.toFixed(2)}`, columns.amount.x, options.pointer.y, { align: 'right' });
}

/**
 * Adds a new page to the PDF document
 *
 * Creates a new page and adds the footer with payment terms.
 *
 * @param doc - The jsPDF document instance
 * @param options - Drawing options with positioning information
 * @param paymentTerms - Payment terms to display in the footer
 */
function addPage(doc: jsPDF, options: DrawOptions, paymentTerms: PaymentTerms) {
  doc.addPage();
  options.pointer.y = options.border.top;
  addFooter(doc, options, paymentTerms);
}

/**
 * Adds footer with payment terms to the PDF document
 *
 * Displays payment terms and banking information at the bottom of the page.
 *
 * @param doc - The jsPDF document instance
 * @param options - Drawing options with positioning information
 * @param paymentTerms - Payment terms and banking information to display
 */
function addFooter(doc: jsPDF, options: DrawOptions, paymentTerms: PaymentTerms): void {
  const xCenter = doc.internal.pageSize.getWidth() / 2;
  let y = options.border.bottom;

  addLine(doc, options);

  y += 6;
  doc.setFont('Helvetica', 'bold');
  doc.setFontSize(options.text.size.tiny);
  doc.setTextColor(options.text.color.lighter);
  doc.setFont('Helvetica', 'normal');
  doc.text(paymentTerms.terms, xCenter, y, { align: 'center' });

  y += 3;
  const bankingText = `BANK: ${paymentTerms.bankName} - ACCOUNT NAME: ${paymentTerms.accountName} - ACCOUNT NO.: ${paymentTerms.accountNumber} - ROUTING NUMBER: ${paymentTerms.routingNumber}`;
  doc.text(bankingText, xCenter, y, { align: 'center' });
}

/**
 * Adds a horizontal line to the PDF document
 *
 * Creates a separator line across the document at the current pointer position.
 *
 * @param doc - The jsPDF document instance
 * @param options - Drawing options with positioning information
 */
function addLine(doc: jsPDF, options: DrawOptions): void {
  doc.setDrawColor(options.line.color.light);
  doc.line(options.border.left, options.pointer.y, options.border.right, options.pointer.y);
}
