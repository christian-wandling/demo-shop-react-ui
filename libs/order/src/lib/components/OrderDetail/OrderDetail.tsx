import { useEffect } from 'react';
import { Link, useParams } from 'react-router';
import { OrderStatusBadge } from '../OrderStatusBadge/OrderStatusBadge';
import { DateTime, LoadingSpinner, NotFound } from '@demo-shop-react-ui/shared';
import { useOrderStore } from '../../+state/useOrderStore';
import { useUserStore } from '@demo-shop-react-ui/user';
import { OrderResponse, UserResponse } from '@demo-shop-react-ui/api';
import { generatePdf } from '../../services/printInvoiceService';

/**
 * Component that displays detailed information about a specific order.
 * Fetches order data based on the ID from the route parameters.
 */
export default function OrderDetail() {
  const { id } = useParams();
  const error = useOrderStore(state => state.error);
  const loading = useOrderStore(state => state.loading);
  useOrderStore(state => state.orders);
  const user = useUserStore(state => state.user);
  const getOrderById = useOrderStore(state => state.getOrderById);
  const fetchOrderById = useOrderStore(state => state.fetchOrderById);

  useEffect(() => {
    if (id) {
      fetchOrderById(+id);
    }
  }, [id, fetchOrderById]);

  if (loading) {
    return <LoadingSpinner />;
  }

  if (!id || error) {
    return <NotFound />;
  }

  const order = getOrderById(+id);

  if (!order || !user) {
    return <NotFound />;
  }

  /**
   * Generates and prints a PDF invoice for the specified order
   *
   * @param order - The order data used to generate the invoice
   * @param user - The user data needed for billing information
   */
  const printPdf = (order: OrderResponse, user: UserResponse) => {
    generatePdf(order, user);
  };

  return (
    <div className="mx-auto mb-8 px-8 max-w-5xl" id="invoice">
      <section className="py-8 top-16 sticky z-10 border-b-2 border-gray-100 bg-white flex justify-between">
        <div className="flex flex-col gap-5">
          <div>
            <h1 className="text-s leading-tight font-bold text-gray-700">
              {user.firstname} {user.lastname}
            </h1>
            <div className="text-xs text-gray-500">
              <p>
                {user.address?.street} {user.address?.apartment}
              </p>
              <p>
                {user.address?.zip} {user.address?.city}, {user.address?.region}
              </p>
              <p>{user.address?.country}</p>
            </div>
          </div>
          <button
            onClick={() => printPdf(order, user)}
            className="btn-print w-36 flex items-center justify-center rounded-md border border-transparent px-8 py-2 text-base font-medium bg-indigo-600 text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2">
            Print
          </button>
        </div>
        <div className="flex flex-col gap-2.5">
          <div className="flex flex-col items-end">
            <div className="text-s leading-tight font-semibold">Order #{order.id}</div>
            <DateTime dateTime={order.created} className="text-xs text-gray-500" />
          </div>
          <OrderStatusBadge status={order.status} />
        </div>
      </section>
      <section>
        <ul className="divide-y divide-gray-100">
          {order.items.map(item => (
            <li key={item.productId} className="flex flex-col sm:flex-row gap-2 sm:gap-6 py-5">
              <Link
                key={item.productId}
                to={`/products/${item.productId}`}
                className="relative w-12 h-12 sm:w-24 sm:h-24 aspect-1 flex-shrink-0 rounded-lg overflow-hidden">
                <img
                  src={item.productThumbnail}
                  alt={item.productName}
                  className="w-full h-full object-cover hover:opacity-75"
                />
              </Link>
              <div className="flex flex-col min-w-0 grow">
                <p className="text-sm sm:leading-6 text-gray-900">
                  #{item.productId} {item.productName}
                </p>
                <p className="sm:mt-1 truncate text-xs sm:leading-5 text-gray-500">Qty {item.quantity}</p>
              </div>
              <div className="flex flex-row sm:flex-col-reverse grow sm:grow-0 justify-between sm:justify-end sm:items-end">
                <p className="sm:mt-1 text-xs leading-5 text-gray-500">
                  <span>
                    {new Intl.NumberFormat('en-US', {
                      style: 'currency',
                      currency: 'USD',
                    }).format(item.unitPrice)}
                  </span>
                </p>
                <p className="text-sm leading-6 text-gray-900">
                  {new Intl.NumberFormat('en-US', { style: 'currency', currency: 'USD' }).format(item.totalPrice)}
                </p>
              </div>
            </li>
          ))}
          <li className="flex justify-between gap-x-6 py-5">
            <div className="flex min-w-0 gap-x-4">
              <div className="min-w-0 flex-auto">
                <p className="text-sm font-semibold leading-6 text-gray-900">total</p>
              </div>
            </div>
            <div className="shrink-0 sm:flex sm:flex-col sm:items-end">
              <p className="text-sm leading-6 text-gray-900">
                {new Intl.NumberFormat('en-US', { style: 'currency', currency: 'USD' }).format(order.amount)}
              </p>
            </div>
          </li>
        </ul>
      </section>
    </div>
  );
}
