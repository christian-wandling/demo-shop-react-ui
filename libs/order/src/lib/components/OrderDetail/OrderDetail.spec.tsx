import { beforeEach, describe, expect, it, vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import OrderDetail from './OrderDetail';
import { useParams } from 'react-router';
import { generatePdf } from '../../services/printInvoiceService';
import { OrderResponse, OrderStatus, UserResponse } from '@demo-shop-react-ui/api';

vi.mock('react-router', () => ({
  useParams: vi.fn(),
  Link: ({ to, children, ...props }: any) => (
    <a href={to} {...props}>
      {children}
    </a>
  ),
}));

const fetchOrderByIdSpy = vi.fn();
const getOrderByIdSpy = vi.fn();

vi.mock('../../+state/useOrderStore', () => ({
  useOrderStore: vi.fn().mockImplementation(selector => {
    const store = {
      fetchOrderById: fetchOrderByIdSpy,
      getOrderById: getOrderByIdSpy,
    };
    return selector(store);
  }),
}));

vi.mock('@demo-shop-react-ui/user', () => ({
  useUserStore: vi.fn().mockReturnValue({
    id: 1,
    firstname: 'John',
    lastname: 'Doe',
    address: {
      street: '123 Main St',
      apartment: 'Apt 4B',
      zip: '12345',
      city: 'New York',
      region: 'NY',
      country: 'USA',
    },
    email: 'email',
    phone: '1234567890',
  }),
}));

vi.mock('../../services/printInvoiceService', () => ({
  generatePdf: vi.fn(),
}));

vi.mock('@demo-shop-react-ui/shared', async importOriginal => ({
  ...(await importOriginal()),
  DateTime: ({ dateTime }: any) => <span data-testid="date-time">{dateTime.toISOString()}</span>,
}));

vi.mock('../OrderStatusBadge/OrderStatusBadge', () => ({
  OrderStatusBadge: ({ status }: any) => <span data-testid="status-badge">{status}</span>,
}));

describe('OrderDetail', () => {
  const mockOrder: OrderResponse = {
    id: 12345,
    userId: 1,
    status: OrderStatus.Created,
    created: new Date('2023-04-01T12:00:00Z'),
    items: [
      {
        productId: 1,
        productName: 'Test Product',
        productThumbnail: '/test-image.jpg',
        quantity: 2,
        unitPrice: 19.99,
        totalPrice: 39.98,
      },
    ],
    amount: 39.98,
  };

  const mockUser: UserResponse = {
    id: 1,
    firstname: 'John',
    lastname: 'Doe',
    address: {
      street: '123 Main St',
      apartment: 'Apt 4B',
      zip: '12345',
      city: 'New York',
      region: 'NY',
      country: 'USA',
    },
    email: 'email',
    phone: '1234567890',
  };

  beforeEach(() => {
    vi.clearAllMocks();
    vi.mocked(useParams).mockReturnValue({ id: '12345' });
    getOrderByIdSpy.mockReturnValue(mockOrder);
  });

  it('fetches order data on mount', () => {
    render(<OrderDetail />);
    expect(fetchOrderByIdSpy).toHaveBeenCalledWith(12345);
  });

  it('matches the snapshot', () => {
    const { baseElement } = render(<OrderDetail />);
    expect(baseElement).toMatchSnapshot();
  });

  it('displays not found page when id is not provided', () => {
    vi.mocked(useParams).mockReturnValue({});
    render(<OrderDetail />);
    expect(screen.getByText('Page not found')).toBeTruthy();
  });

  it('displays page not found when order or user data is not available', () => {
    fetchOrderByIdSpy.mockReturnValue(null);
    getOrderByIdSpy.mockReturnValue(undefined);
    render(<OrderDetail />);
    expect(screen.getByText('Page not found')).toBeTruthy();
  });

  it('displays order details correctly when data is available', () => {
    render(<OrderDetail />);

    expect(screen.getByText('John Doe')).toBeTruthy();
    expect(screen.getByText('123 Main St Apt 4B')).toBeTruthy();
    expect(screen.getByText('12345 New York, NY')).toBeTruthy();
    expect(screen.getByText('USA')).toBeTruthy();

    expect(screen.getByText('Order #12345')).toBeTruthy();
    expect(screen.getByTestId('date-time')).toBeTruthy();
    expect(screen.getByTestId('status-badge')).toBeTruthy();

    expect(screen.getByText('#1 Test Product')).toBeTruthy();
    expect(screen.getByText('Qty 2')).toBeTruthy();
    expect((screen.getByAltText('Test Product') as HTMLImageElement).src).toMatch(/test-image.jpg/);

    const priceElements = screen.getAllByText(/\$\d+\.\d+/);
    expect(priceElements.length).toBeGreaterThan(0);
  });

  it('calls generatePdf when "Print" button is clicked', () => {
    render(<OrderDetail />);
    const printButton = screen.getByText('Print');

    fireEvent.click(printButton);

    expect(generatePdf).toHaveBeenCalledWith(mockOrder, mockUser);
  });

  it('links to product detail page correctly', () => {
    render(<OrderDetail />);
    const productLink: HTMLLinkElement = screen.getByRole('link');

    expect(productLink.href).toMatch(/products\/1/);
  });
});
