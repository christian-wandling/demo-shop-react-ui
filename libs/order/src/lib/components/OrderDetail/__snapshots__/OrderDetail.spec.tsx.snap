// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OrderDetail > matches the snapshot 1`] = `
<body>
  <div>
    <div
      class="mx-auto mb-8 px-8 max-w-5xl"
      id="invoice"
    >
      <section
        class="py-8 top-16 sticky z-10 border-b-2 border-gray-100 bg-white flex justify-between"
      >
        <div
          class="flex flex-col gap-5"
        >
          <div>
            <h1
              class="text-s leading-tight font-bold text-gray-700"
            >
              John
               
              Doe
            </h1>
            <div
              class="text-xs text-gray-500"
            >
              <p>
                123 Main St
                 
                Apt 4B
              </p>
              <p>
                12345
                 
                New York
                , 
                NY
              </p>
              <p>
                USA
              </p>
            </div>
          </div>
          <button
            class="btn-print w-36 flex items-center justify-center rounded-md border border-transparent px-8 py-2 text-base font-medium bg-indigo-600 text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2"
          >
            Print
          </button>
        </div>
        <div
          class="flex flex-col gap-2.5"
        >
          <div
            class="flex flex-col items-end"
          >
            <div
              class="text-s leading-tight font-semibold"
            >
              Order #
              12345
            </div>
            <span
              data-testid="date-time"
            >
              2023-04-01T12:00:00.000Z
            </span>
          </div>
          <span
            data-testid="status-badge"
          >
            Created
          </span>
        </div>
      </section>
      <section>
        <ul
          class="divide-y divide-gray-100"
        >
          <li
            class="flex flex-col sm:flex-row gap-2 sm:gap-6 py-5"
          >
            <a
              class="relative w-12 h-12 sm:w-24 sm:h-24 aspect-1 flex-shrink-0 rounded-lg overflow-hidden"
              href="/products/1"
            >
              <img
                alt="Test Product"
                class="w-full h-full object-cover hover:opacity-75"
                src="/test-image.jpg"
              />
            </a>
            <div
              class="flex flex-col min-w-0 grow"
            >
              <p
                class="text-sm sm:leading-6 text-gray-900"
              >
                #
                1
                 
                Test Product
              </p>
              <p
                class="sm:mt-1 truncate text-xs sm:leading-5 text-gray-500"
              >
                Qty 
                2
              </p>
            </div>
            <div
              class="flex flex-row sm:flex-col-reverse grow sm:grow-0 justify-between sm:justify-end sm:items-end"
            >
              <p
                class="sm:mt-1 text-xs leading-5 text-gray-500"
              >
                <span>
                  $19.99
                </span>
              </p>
              <p
                class="text-sm leading-6 text-gray-900"
              >
                $39.98
              </p>
            </div>
          </li>
          <li
            class="flex justify-between gap-x-6 py-5"
          >
            <div
              class="flex min-w-0 gap-x-4"
            >
              <div
                class="min-w-0 flex-auto"
              >
                <p
                  class="text-sm font-semibold leading-6 text-gray-900"
                >
                  total
                </p>
              </div>
            </div>
            <div
              class="shrink-0 sm:flex sm:flex-col sm:items-end"
            >
              <p
                class="text-sm leading-6 text-gray-900"
              >
                $39.98
              </p>
            </div>
          </li>
        </ul>
      </section>
    </div>
  </div>
</body>
`;
